app/scripts/background.js

```diff
@@ -91,16 +92,60 @@ const PHISHING_WARNING_PAGE_TIMEOUT = ONE_SECOND_IN_MILLISECONDS;
 /**
  * In case of MV3 we attach a "onConnect" event listener as soon as the application is initialised.
  * Reason is that in case of MV3 a delay in doing this was resulting in missing first connect event after service worker is re-activated.
+ *
+ * @param remotePort
  */
-
 const initApp = async (remotePort) => {
   browser.runtime.onConnect.removeListener(initApp);
   await initialize(remotePort);
   log.info('MetaMask initialization complete.');
 };
 
+const throwErrorIfLastErrorFound = () => {
+  const err = checkForError();
+  if (err) {
+    throw err;
+  }
+};
+
+/**
+ * Sends a message to the dapp(s) content script to signal it can connect to MetaMask background as
+ * the backend is not active. It is required to re-connect dapps after service worker re-activates.
+ * For non-dapp pages, the message will be sent and ignored.
+ */
+const sendReadyMessageToTabs = async () => {
+  try {
+    const tabs = await browser.tabs.query({});
+    throwErrorIfLastErrorFound();
+
+    /* TODO we should only sendMessage to dapp tabs, not all tabs. */
```

Comment:
Could this (sending to all tabs unconditionally) be a metadata leak (indicating that the user has MetaMask installed and active)?

Reply:
@legobeat  reply for this comment and https://github.com/MetaMask/metamask-extension/pull/16075#discussion_r1012370017. Good question

I believe that we're not leaking any metadata here. Content scripts run in an [isolated world](https://developer.chrome.com/docs/extensions/mv3/content_scripts/#isolated_world). 
> An isolated world is a private execution environment that isn't accessible to the page or other extensions.

Even if we are sending messages to all MetaMask content scripts across all pages, the pages themselves cannot manually read the data or messages from the content scripts. 

Some other things to note: 
- our content script does not accept messages from external extensions or from web pages
- our content script does not send messages outside of our extension or channels

cc: @Gudahtt @jpuri

Reply:
Very educative, thank you!

Reply:
>indicating to the page on any tab that the user has MetaMask installed and active

We do currently leak that a user has MetaMask installed, unfortunately. It's a consequence of our dapp API, which uses an injected provider (the `window.ethereum` object). We are considering moving away from that pattern, but it'll be something we'll have to support for a very long time for backwards compatibility even if we start moving away from it.

But luckily no details of the users accounts are exposed by default.

Reply:
Indeed unfortunate. *If* this would have been a leak, it would provide additional fingerprinting entropy and be a proxy for user activity and whether the user is an *active* MM user. Fortunately that does not seem to be a concern here. +1 on a more privacy-respecting future API.

Reply:
ah yes. the injected provider does leak that the user has installed MetaMask. I meant from this specific code, we are not leaking the data. sorry, I should have clarified this earlier.

Thanks for checking in on these concerns and sharing your thoughts! @legobeat 

> +1 on a more privacy-respecting future API

+1